src/main/java/components/Parser.java

```diff
@@ -0,0 +1,194 @@
+package components;
+
+import java.time.LocalDate;
+import java.time.format.DateTimeParseException;
+
+/**
+ * Represents a parser, which tries to understand user input.
+ */
+public class Parser {
+
+  private static Ui Ui;
+  private static TaskList taskList;
+  private static Storage storage;
+
+  /**
+   * Connects Ui object with Parser.
+   *
+   * @param ui Ui to be connected with Parser.
+   */
+  public static void setUi(Ui ui) {
+    Parser.Ui = ui;
+  }
+
+  /**
+   * Connects TaskList object with Parser.
+   *
+   * @param taskList taskList to be connected with Parser.
+   */
+  public static void setTaskList(TaskList taskList) {
+    Parser.taskList = taskList;
+  }
+
+  /**
+   * Parses and trying to make sense of user input.
+   *
+   * @param line string representation of user input.
+   * @throws DukeException If user input is incoherent.
+   */
+  public static void parseLine(String line) throws DukeException {
+    if (line.equals("bye")) {
+      System.out.println("Bye. Hope to see you again soon!");
+      System.exit(0);
+    } else if (line.equals("list")) {
+      taskList.showTasks();
+    } else if (line.contains("unmark")) {
+      if (line.equals("unmark")) {
+        throw new DukeException("☹ OOPS!!! The description of a mark cannot be empty.");
```

Comment:
Maybe we could edit the exception so we know where the error occurs? I.e. unmark instead of mark. Nice validation of inputs tho, keep it up! 😃